### PR TITLE
sam0_common: Use Single-cycle I/O Port for GPIO when available

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

The Cortex-m0 based ATSAM devices can use the Single-cycle I/O Port for GPIO. This commit modifies the gpio_t type to use this port when available. It is mapped back to the peripheral memory space for configuration access.  When it is not available, the _port_iobus() and _port() functions behave identical, which is the case for the samd51.

The possible downside of this PR is that continuous sampling is required for GPIO IN pins. I have no idea what the impact is on power consumption and the reference manual doesn't mention it.

### Testing procedure

Test both GPIO read and write on the samr21-xpro (or similar) and on the same54-xpro. Using SAUL on the `examples/default` is sufficient for this.

### Benchmarks

`tests/periph_gpio`:

<details><summary>Master</summary>

```
bench 1 0
2020-08-20 12:23:28,723 # bench 1 0
2020-08-20 12:23:28,724 # 
2020-08-20 12:23:28,727 # GPIO driver run-time performance benchmark
2020-08-20 12:23:28,727 # 
2020-08-20 12:23:28,750 #                  nop loop:     14592us  ---   0.145us per call  ---    6853070 calls per sec
2020-08-20 12:23:28,804 #                  gpio_set:     45842us  ---   0.458us per call  ---    2181405 calls per sec
2020-08-20 12:23:28,864 #                gpio_clear:     52092us  ---   0.520us per call  ---    1919680 calls per sec
2020-08-20 12:23:28,918 #               gpio_toggle:     45842us  ---   0.458us per call  ---    2181405 calls per sec
2020-08-20 12:23:29,006 #                 gpio_read:     79176us  ---   0.791us per call  ---    1263008 calls per sec
2020-08-20 12:23:29,083 #                gpio_write:     68759us  ---   0.687us per call  ---    1454355 calls per sec
2020-08-20 12:23:29,084 # 
2020-08-20 12:23:29,085 #  --- DONE ---
```

</details>

<details><summary>This PR</summary>

```
bench 0 0
2020-08-20 12:22:58,858 # bench 0 0
2020-08-20 12:22:58,859 # 
2020-08-20 12:22:58,862 # GPIO driver run-time performance benchmark
2020-08-20 12:22:58,862 # 
2020-08-20 12:22:58,884 #                  nop loop:     14592us  ---   0.145us per call  ---    6853070 calls per sec
2020-08-20 12:22:58,932 #                  gpio_set:     39592us  ---   0.395us per call  ---    2525762 calls per sec
2020-08-20 12:22:58,984 #                gpio_clear:     43759us  ---   0.437us per call  ---    2285244 calls per sec
2020-08-20 12:22:59,032 #               gpio_toggle:     39592us  ---   0.395us per call  ---    2525762 calls per sec
2020-08-20 12:22:59,109 #                 gpio_read:     68759us  ---   0.687us per call  ---    1454355 calls per sec
2020-08-20 12:22:59,182 #                gpio_write:     64592us  ---   0.645us per call  ---    1548179 calls per sec
2020-08-20 12:22:59,183 # 
2020-08-20 12:22:59,183 #  --- DONE ---
```

</details>

Or about 3 - 5 CPU cycles per call.

### Issues/PRs references

None

### CODE REVIEW

**Comment by bergzand** [comment]
@gschorcht Does this PR conflict with your GPIO rework efforts?

File: cpu/sam0_common/periph/gpio.c [hunk 1]
```diff
@@ -75,11 +75,25 @@ typedef enum {
...
+ */
+static inline PortGroup *_port_iobus(gpio_t pin)
 {
     return (PortGroup *)(pin & ~(0x1f));
 }
 
+static inline PortGroup *_port(gpio_t pin)
+{
+#ifdef PORT_IOBUS
+    return (PortGroup *)((pin & ~(0x1f)) - (uintptr_t)PORT_IOBUS + (uintptr_t)PORT);
```
Comment (benpicco) [inline cpu/sam0_common/periph/gpio.c:91]:
```suggestion
    return (PortGroup *)((uintptr_t)_port_iobus(pin) - (uintptr_t)PORT_IOBUS + (uintptr_t)PORT);
```

**Review by benpicco — COMMENTED** [review]
> It is mapped back to the peripheral memory space for configuration access.

where does this happen?

**Comment by bergzand** [comment]
> where does this happen?

Added a comment here:

https://github.com/RIOT-OS/RIOT/blob/f8e9f2749671c5e4c31fef8b7df112e6d1ed14a0/cpu/sam0_common/periph/gpio.c#L90-L95

File: cpu/sam0_common/periph/gpio.c (hunk 1 above)
Comment (bergzand) [reply cpu/sam0_common/periph/gpio.c]: Yup, fixed

**Comment by benpicco** [comment]
Please squash.

**Comment by bergzand** [comment]
Squashed!

**Comment by benpicco** [comment]
On `saml10-xpro` I connected PA16 and PA17 with a jumper (after reverting 43d6c4147a8034a1a115a061a1b8ca0705c86222)

```
2020-08-20 13:14:37,761 #  auto_test 0 16 0 17
2020-08-20 13:14:37,762 # [START]
2020-08-20 13:14:37,766 # gpio_set() or gpio_read() or failed
```

on `master`:

```
2020-08-20 13:15:00,995 #  auto_test 0 16 0 17
2020-08-20 13:15:00,996 # [START]
2020-08-20 13:15:01,003 # [SUCCESS]
```

**Comment by bergzand** [comment]
> On `saml10-xpro` I connected PA16 and PA17 with a jumper (after reverting [43d6c41](https://github.com/RIOT-OS/RIOT/commit/43d6c4147a8034a1a115a061a1b8ca0705c86222))

Thanks for testing. I managed to remove the code enabling/disabling continuous sampling for the GPIO pins in one of my pushes. I've restored it and now the `auto_test` works for me on the samr21-xpro:

```
2020-08-20 13:22:01,651 # auto_test 0 16 0 17
2020-08-20 13:22:01,652 # [START]
2020-08-20 13:22:01,659 # [SUCCESS]
```

**Comment by benpicco** [comment]
Also works on `saml10-xpro` now

```
2020-08-20 13:24:33,256 #  auto_test 0 16 0 17
2020-08-20 13:24:33,257 # [START]
2020-08-20 13:24:33,264 # [SUCCESS]
```

**Comment by bergzand** [comment]
Awesome, thanks for measuring :)

I can restore the original code for the `gpio_read()` case and only use this for writes. What do you think?

**Comment by benpicco** [comment]
Would be interesting to see the effects of that.
Another option would be to let the user select between performance / power efficiency at compile time.

**Comment by bergzand** [comment]
> Would be interesting to see the effects of that.

Using a regular `_port(pin);` for `gpio_read()`:

```
2020-08-20 17:33:47,803 #                  nop loop:     14592us  ---   0.145us per call  ---    6853070 calls per sec
2020-08-20 17:33:47,850 #                  gpio_set:     39592us  ---   0.395us per call  ---    2525762 calls per sec
2020-08-20 17:33:47,902 #                gpio_clear:     43759us  ---   0.437us per call  ---    2285244 calls per sec
2020-08-20 17:33:47,950 #               gpio_toggle:     39592us  ---   0.395us per call  ---    2525762 calls per sec
2020-08-20 17:33:48,050 #                 gpio_read:     91676us  ---   0.916us per call  ---    1090798 calls per sec
2020-08-20 17:33:48,123 #                gpio_write:     64592us  ---   0.645us per call  ---    1548179 calls per sec
```

**Comment by benpicco** [comment]
How is the effect on power consumption?

**Comment by bergzand** [comment]
> How is the effect on power consumption?

I don't have the gear here to measure the impact. Do you mind measuring the latest version here another time.



>  Another option would be to let the user select between performance / power efficiency at compile time.

I think this is the best option here, let the user select it based on the requirements.

**Comment by benpicco** [comment]
> I don't have the gear here to measure the impact. 

The `-xpro` boards all have a jumper that bridges Vcc MCU with Vcc Target.
If you replace the jumper with an ammeter you can measure the current draw.

I'll try that later. Do you have a patch that only uses the new path for output GPIOs so I can directly compare the two?

**Comment by bergzand** [comment]
> The `-xpro` boards all have a jumper that bridges Vcc MCU with Vcc Target.

The samr21-xpro I have here always measures around 10mA, which I think is caused by the edbg status LED's :(

**Comment by bergzand** [comment]
> I'll try that later. Do you have a patch that only uses the new path for output GPIOs so I can directly compare the two?

Guess I forgot to push here :(

**Comment by bergzand** [comment]
> I'll try that later. Do you have a patch that only uses the new path for output GPIOs so I can directly compare the two?

See 9d1b60f

**Comment by benpicco** [comment]
Yup that did the trick.
Power consumption is now the same as on `master`.

**Review by benpicco — APPROVED** [review]
`gpio_write()` got slightly better, `gpio_read()` got slightly worse.
No power regressions and no influence on hardware without this feature.

**Comment by dylad** [comment]
Sorry to jump this late into this PR !
I noticed in SAMD21 and SAML21 datasheet that continuous sampling should be enable for IN registers when using IOBUS.
Maybe this can improve the `gpio_read()` situation ?

**Comment by benpicco** [comment]
> I noticed in SAMD21 and SAML21 datasheet that continuous sampling should be enable for IN registers when using IOBUS.
> Maybe this can improve the `gpio_read()` situation ?

It will and that was what 26a08606bccb694dbcefd4820185aa9aa45066b7 initially did.
However, that would increase power consumption (https://github.com/RIOT-OS/RIOT/pull/14804#issuecomment-677565310) on these low-power parts.
And when people select these chips they will likely do so because of power consumption, not because of I/O performance - so we shouldn't make this a default.

I'd be fine with a config option if it doesn't make the code messy.

**Comment by dylad** [comment]
> - so we shouldn't make this a default.

Not making it the default choice is definitely the way to go.

> However, that would increase power consumption (#14804 (comment)) on these low-power parts.
And when people select these chips they will likely do so because of power consumption, not because of I/O performance

This is a necessary trade off if users really want to use this feature for whatever reasons. We should just advertise it.

> I'd be fine with a config option if it doesn't make the code messy.

+1

**Comment by benpicco** [comment]
How about this patch

```patch
--- a/cpu/sam0_common/periph/gpio.c
+++ b/cpu/sam0_common/periph/gpio.c
@@ -139,9 +139,15 @@ int gpio_init(gpio_t pin, gpio_mode_t mode)
 
     /* set pin direction */
     if (mode & 0x2) {
+        if (IS_ACTIVE(PERIPH_GPIO_FAST_READ)) {
+            port->CTRL.reg |= pin_mask;
+        }
         port->DIRCLR.reg = pin_mask;
     }
     else {
+        if (IS_ACTIVE(PERIPH_GPIO_FAST_READ)) {
+            port->CTRL.reg &= ~pin_mask;
+        }
         port->DIRSET.reg = pin_mask;
     }
 
@@ -161,9 +167,16 @@ int gpio_init(gpio_t pin, gpio_mode_t mode)
 
 int gpio_read(gpio_t pin)
 {
-    PortGroup *port = _port(pin);
+    PortGroup *port;
     int mask = _pin_mask(pin);
 
+    if (IS_ACTIVE(PERIPH_GPIO_FAST_READ)) {
+        port = _port_iobus(pin);
+    }
+    else {
+        port = _port(pin);
+    }
+
     if (port->DIR.reg & mask) {
         return (port->OUT.reg & mask) ? 1 : 0;
     }
```

This way we can enable the feature with the `periph_gpio_fast_read` pseudo-module / feature.

**Comment by bergzand** [comment]
> This way we can enable the feature with the `periph_gpio_fast_read` pseudo-module / feature.

But this is a `periph_gpio_fast_write` :D

**Comment by benpicco** [comment]
I thought we can always enable the fast write, but the fast read will cost us more power.

**Comment by bergzand** [comment]
> I thought we can always enable the fast write, but the fast read will cost us more power.

Oh, yeah, totally fine with me.

**Comment by dylad** [comment]
> I thought we can always enable the fast write, but the fast read will cost us more power.

This is fine by me too :)

**Comment by bergzand** [comment]
> ```diff
> +        if (IS_ACTIVE(PERIPH_GPIO_FAST_READ)) {
> ```

Almost perfect :)

File: cpu/sam0_common/periph/gpio.c [hunk 2]
```diff
@@ -21,6 +21,18 @@
  * @author      Kaspar Schleiser <kaspar@schleiser.de>
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
  *
+ * The GPIO implementation here support the PERIPH_GPIO_FAST_READ pseudomodule
+ * on the cortex-m0 based devices.  This trades an increase in power consumption
+ * for a decrease in GPIO pin read latency. Enable this with:
+ *
+ * ```
+ * USEMODULE += periph_gpio_fast_read
```
Comment (benpicco) [inline cpu/sam0_common/periph/gpio.c:29]: You want to add `FEATURES_PROVIDED += periph_gpio_fast_read` to `saml1x` and `samd21`.
(Don't forget Kconfig :wink:)

Then we can simply do 

    FEATURES_OPTIONAL += periph_gpio_fast_read

(I'd add that to `tests/periph_gpio` so this code path is tested)

File: tests/periph_gpio/Makefile [hunk 3]
```diff
@@ -2,11 +2,11 @@ include ../Makefile.tests_common
 
 FEATURES_REQUIRED = periph_gpio
 FEATURES_OPTIONAL = periph_gpio_irq
+FEATURES_OPTIONAL = periph_gpio_fast_read
```
Comment (benpicco) [inline tests/periph_gpio/Makefile:5]:
```suggestion
FEATURES_OPTIONAL += periph_gpio_fast_read
```

File: cpu/sam0_common/periph/gpio.c (hunk 2 above)
Comment (bergzand) [reply cpu/sam0_common/periph/gpio.c]: Makes sense, let me know if I forgot one of the sam0 variants

File: tests/periph_gpio/Makefile [hunk 4]
```diff
@@ -2,11 +2,11 @@ include ../Makefile.tests_common
...
 
 USEMODULE += shell
 USEMODULE += shell_commands
 USEMODULE += benchmark
-
```
Comment (benpicco) [inline tests/periph_gpio/Makefile]: I'd leave the newline

File: cpu/samd21/Kconfig [hunk 5]
```diff
@@ -10,6 +10,7 @@ config CPU_FAM_SAMD21
     select CPU_COMMON_SAM0
     select CPU_CORE_CORTEX_M0PLUS
     select HAS_CPU_SAMD21
+    select HAS_PERIPH_GPIO_FAST_READ
```
Comment (benpicco) [inline cpu/samd21/Kconfig:13]: this needs to be defined in `kconfigs/Kconfig.features`

File: cpu/sam0_common/periph/gpio.c (hunk 2 above)
Comment (benpicco) [reply cpu/sam0_common/periph/gpio.c]: Nope, I would have forgotten about saml21, but it's a cm0+ too!
samd5x does not have this feature.

File: tests/periph_gpio/Makefile (hunk 3 above)
Comment (bergzand) [reply tests/periph_gpio/Makefile]: Good catch!

File: tests/periph_gpio/Makefile (hunk 4 above)
Comment (bergzand) [reply tests/periph_gpio/Makefile]: Already fixed

File: cpu/samd21/Kconfig (hunk 5 above)
Comment (bergzand) [reply cpu/samd21/Kconfig:13]: I was under the impression that I was just reusing a feature already used by a different MCU. Now I feel betrayed :disappointed: 

Added it to the feature list with a generic description.

**Comment by benpicco** [comment]
Please squash!

File: cpu/sam0_common/periph/gpio.c [hunk 6]
```diff
@@ -21,6 +21,18 @@
  * @author      Kaspar Schleiser <kaspar@schleiser.de>
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
  *
+ * The GPIO implementation here support the PERIPH_GPIO_FAST_READ pseudomodule
+ * on the cortex-m0 based devices.  This trades an increase in power consumption
+ * for a decrease in GPIO pin read latency. Enable this in your makefile with:
+ *
+ * ```
+ * FEATURES_REQUIRED += periph_gpio_fast_read
```
Comment (benpicco) [inline cpu/sam0_common/periph/gpio.c:29]: `FEATURES_OPTIONAL` would also be enough

**Comment by bergzand** [comment]
Squashed!

File: cpu/sam0_common/periph/gpio.c (hunk 2 above)
Comment (bergzand) [reply cpu/sam0_common/periph/gpio.c]: I've added the cortex-m23 to the docs here

**Comment by bergzand** [comment]
Thanks for reviewing!

**Pull request merged**